serve/mlc_serve/engine/engine_common.py

```diff
@@ -132,6 +133,31 @@ def detokenize_incrementally(
     return delta
 
 
+def logprob_detokenize(tokenizer: TokenizerP, logprob_info: Optional[LOGPROBS_TYPE]) -> Optional[LOGPROBS_TYPE]:
+    """Detokenize logprob information"""
+    if logprob_info is None:
+        return None
+    (res, res_logprob), top_tokens = logprob_info
+    top_tokens = list(top_tokens)
+    count: Dict[str, int] = {}
+    top_logprobs: List[Tuple] = []
+    # dedup duplicates
+    # Todo: Make sure decode can generate different tokens
+    for top_token, _ in top_tokens:
+        detokenized = tokenizer.decode(top_token)
+        if detokenized in count:
+            count[detokenized] += 1
+        else:
+            count[detokenized] = 1
+    for top_token, top_logprob in top_tokens:
+        detokenized = tokenizer.decode(top_token)
+        if count[detokenized] == 1:
+            top_logprobs.append((detokenized, float(top_logprob)))
+        else:
+            top_logprobs.append((f"{detokenized}_{top_token}", float(top_logprob)))
```

Comment:
These two for loops look really slow depending on the length of `top_tokens`. Also please document what they are for more clearly.

Reply:
For some reason different tokens may share the same detokenized results if I use detokenizer to detokenize them. For example, see the following response:
```
          {
            "token": "merge",
            "logprob": -0.03087696246802807,
            "bytes": null,
            "top_logprobs": [
              {
                "token": "merge_10366",
                "logprob": -0.03087696246802807,
                "bytes": null
              },
              {
                "token": "Mer_4702",
                "logprob": -3.796501874923706,
                "bytes": null
              },
              {
                "token": "mer",
                "logprob": -5.093377113342285,
                "bytes": null
              },
              {
                "token": "merge_14634",
                "logprob": -6.702752113342285,
                "bytes": null
              },
              {
                "token": "Mer_15836",
                "logprob": -7.984002113342285,
                "bytes": null
              }
            ]
          }
```
Token `4702` and `15836` can both be tokenized as `Mer`. I think it's due to some defect of the tokenizer. We can skip the loops if we don't care about the duplications caused by this issue.

Reply:
Detokenizer generally needs more than one token to work correctly. For example, during each decoding step we detokenize the whole prompt + generated tokens so far, and only pick up the postfix delta.